airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/CatalogHelpers.java

```diff
@@ -124,31 +134,151 @@ public static Set<String> getTopLevelFieldNames(final ConfiguredAirbyteStream st
+   * are not returned.
+   *
+   * @param jsonSchema - a JSONSchema node
+   * @return a list of all keys for all objects within the node. ordered in depth-first search
+   *         preorder.
+   */
+  @VisibleForTesting
+  protected static List<Pair<List<String>, JsonNode>> getFullyQualifiedFieldNamesWithTypes(final JsonNode jsonSchema) {
+    // if this were ever a performance issue, it could be replaced with a trie. this seems unlikely
+    // however.
+    final Set<List<String>> fieldNamesThatAreOneOfs = new HashSet<>();
+
+    return JsonSchemas.traverseJsonSchemaWithCollector(jsonSchema, (node, basicPath) -> {
+      final List<String> fieldName = basicPath.stream().filter(fieldOrList -> !fieldOrList.isList()).map(FieldNameOrList::getFieldName).toList();
+      return Pair.of(fieldName, node);
+    })
+        .stream()
+        // first node is the original object.
+        .skip(1)
+        .filter(fieldWithSchema -> filterChildrenOfFoneOneOf(fieldWithSchema.getLeft(), fieldWithSchema.getRight(), fieldNamesThatAreOneOfs))
+        .toList();
+  }
+
+  /**
+   * Predicate that checks if a field is a CHILD of a oneOf field. If child of a oneOf, returns false.
+   * Otherwise, true. This method as side effects. It assumes that it will be run in order on field
+   * names returned in depth-first search preoorder. As it encounters oneOfs it adds them to a
+   * collection. It then checks if subsequent field names are prefix matches to the field that are
+   * oneOfs.
+   *
+   * @param fieldName - field to investigate
+   * @param schema - schema of field
+   * @param oneOfFieldNameAccumulator - collection of fields that are oneOfs
+   * @return If child of a oneOf, returns false. Otherwise, true.
+   */
+  private static boolean filterChildrenOfFoneOneOf(final List<String> fieldName,
+                                                   final JsonNode schema,
+                                                   final Set<List<String>> oneOfFieldNameAccumulator) {
+    if (isOneOfField(schema)) {
+      oneOfFieldNameAccumulator.add(fieldName);
+      // return early because we know it is a oneOf and therefore cannot be a child of a oneOf.
+      return true;
+    }
+
+    // leverage that nodes are returned in depth-first search preorder. this means the parent field for
+    // the oneOf will be present in the list BEFORE any of its children.
+    for (final List<String> oneOfFieldName : oneOfFieldNameAccumulator) {
+      final String oneOfFieldNameString = String.join(".", oneOfFieldName);
+      final String fieldNameString = String.join(".", fieldName);
+
+      if (fieldNameString.startsWith(oneOfFieldNameString)) {
+        return false;
       }
     }
+    return true;
+  }
 
-    return allFieldNames;
+  private static boolean isOneOfField(final JsonNode schema) {
+    return !MoreIterators.toSet(schema.fieldNames()).contains("type");
```

Comment:
This will also be true for anyOf and allOf is it some field we want to filter as well? Or is it something not supported in the catalog.